File: src/providers/openai-cua/rules/set-safety-identifier.ts
/**
 * Flags a `responses.create()` call with no `safety_identifier` (or the
 * older `user`) parameter — without one, OpenAI attributes policy
 * violations to the whole shared API key rather than a single end user.
 */
import { isResponsesCreateCall } from '../utils.js';

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'responses.create() must set safety_identifier for per-end-user policy attribution',
      category: 'integration',
      rationale:
        'OpenAI documents that a stable per-end-user safety_identifier lets policy violations be attributed and acted on per end user. Without one on a multi-tenant integration routing many customers through a single shared API key, a single customer triggering a high-confidence policy-violation heuristic can result in access being temporarily revoked for the entire organization, not just the offending identifier.',
      docsUrl: 'https://help.openai.com/en/articles/5428082-how-to-incorporate-a-safety-identifier',
      recommended: true,
    },
    messages: {
      missingSafetyIdentifier:
        'This responses.create() call sets no safety_identifier (or user) — a policy violation here could be attributed to the entire shared API key instead of one end user.',
    },
  },
  create(context: any) {
    function propName(node: any): string | undefined {
      if (!node) return undefined;
      if (node.type === 'Identifier') return node.name;
      if (node.type === 'Literal' && typeof node.value === 'string') return node.value;
      return undefined;
    }

    return {
      CallExpression(node: any) {
        if (!isResponsesCreateCall(node)) return;

        const optionsArg = node.arguments?.[0];
        if (optionsArg?.type !== 'ObjectExpression') return;

        const hasIdentifier = (optionsArg.properties ?? []).some((p: any) => {
          if (p?.type !== 'Property') return false;
          const keyName = propName(p.key);
          return keyName === 'safety_identifier' || keyName === 'user';
        });

        if (!hasIdentifier) {
          context.report({ node, messageId: 'missingSafetyIdentifier' });
        }
      },
    };
  },
};

export const openaiCuaSetSafetyIdentifierRule = rule;

File: src/providers/openai-cua/rules/python/set_safety_identifier.py
"""openai-cua-set-safety-identifier

Parity with the JS rule: `responses.create()` with no `safety_identifier` (or
the older `user`) keyword argument attributes policy violations to the whole
shared API key rather than a single end user.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import is_responses_create_call, kwarg_names  # noqa: E402

RULE_KEY = "openai-cua-set-safety-identifier"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or not is_responses_create_call(node):
            continue
        names = kwarg_names(node)
        if "safety_identifier" in names or "user" in names:
            continue
        # `**kwargs` unpacking — can't verify either way, don't flag.
        if any(k.arg is None for k in node.keywords):
            continue
        line, col, _, _ = loc(node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})
    return out
